### PR TITLE
test: add initial integration test scripts

### PR DESCRIPTION
**Pull request by jpeach**
Add an initial set of integration test documents, along with
helper scripts to install a kind cluster and to run the tests.

This updates #2223.

Signed-off-by: James Peach <jpeach@vmware.com>

### CODE REVIEW

**Comment by jpeach** [comment]
Test documents and preliminary scripts. Next steps are integration into the build (not into `make check` though) and testing out what we can do in CI.

File: _integration/testsuite/fixtures/httpbin.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    matchLabels:
+      app.kubernetes.io/name: *name
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: *name
+    spec:
+      containers:
+      - name: httpbin
+        image: docker.io/kennethreitz/httpbin
```
Comment (stevesloka) [inline _integration/testsuite/fixtures/httpbin.yaml:31]: Should we pin to tags and not latest?

File: _integration/testsuite/fixtures/ingress-conformance-echo.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    matchLabels:
+      app.kubernetes.io/name: *name
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: *name
+    spec:
+      containers:
+      - name: conformance-echo
+        image: agervais/ingress-conformance-echo:latest
```
Comment (stevesloka) [inline _integration/testsuite/fixtures/ingress-conformance-echo.yaml:31]: latest?

File: _integration/testsuite/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Contour integration tests
+
+For now, run integration tests manually.
+
+The [make-kind-cluster.sh](./make-kind-cluster.sh) script will bring up
+a local kind cluster. It will build Contour from the working repository
+and install it, along with [cert-manager](https://cert-manager.io),
+which is needed for tests that use TLS.
+
+You will need to install the [Integration Tester for Kubernetes](https://github.com/projectcontour/integration-tester).
```
Comment (stevesloka) [inline _integration/testsuite/README.md:10]: I was going to test this, but there's nothing to install just yet? https://github.com/projectcontour/integration-tester

File: _integration/testsuite/httpproxy/004-https-sni-enforcement.yaml [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import data.contour.resources
+
+# Ensure that cert-manager is installed.
+# Version check the certificates resource.
+
+Group := "cert-manager.io"
```
Comment (stevesloka) [inline _integration/testsuite/httpproxy/004-https-sni-enforcement.yaml:20]: Should this stuff be in a generic top level set of code? I assume we'll need cert-manager in this example in other tests. How do we keep all this in sync?

File: _integration/testsuite/README.md (hunk 3 above)
Comment (stevesloka) [reply _integration/testsuite/README.md:10]: I see there's a PR for that.

File: _integration/testsuite/make-kind-cluster.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        "$@"
+}
+
+if kind::cluster::exists "$CLUSTERNAME" ; then
+    echo "cluster $CLUSTERNAME already exists"
+    exit 2
+fi
+
+# Build the current version of Contour.
+make -C ${REPO} container IMAGE=docker.io/projectcontour/contour VERSION="v$$"
```
Comment (stevesloka) [inline _integration/testsuite/make-kind-cluster.sh:53]: To me, building a kind cluster and then building an image should be separate things. What if I want to use an existing cluster and skip that step?

File: _integration/testsuite/README.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+To run the tests, use the [run-test-case.sh](./run-test-case.sh)
+script. The test output can be verbose, so prefer to run one at a time.
+
+The tests for for the HTTPProxy API are in the [httpproxy](./httpproxy)
```
Comment (stevesloka) [inline _integration/testsuite/README.md:15]: nit: `for for`

File: _integration/testsuite/README.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+For now, run integration tests manually.
+
+The [make-kind-cluster.sh](./make-kind-cluster.sh) script will bring up
+a local kind cluster. It will build Contour from the working repository
+and install it, along with [cert-manager](https://cert-manager.io),
+which is needed for tests that use TLS.
+
+You will need to install the [Integration Tester for Kubernetes](https://github.com/projectcontour/integration-tester).
+
+To run the tests, use the [run-test-case.sh](./run-test-case.sh)
```
Comment (stevesloka) [inline _integration/testsuite/README.md:12]: How do we just run them all? How do you know feedback as to if it passed or failed?

I ran this test, but not sure if it's good or not:

```
 $ ./run-test-case.sh httpproxy/001-required-field-validation.yaml                                                                         [16:02:48]
TAP version 13
# reading document from httpproxy/001-required-field-validation.yaml
# decoding document with 4 parts from httpproxy/001-required-field-validation.yaml
# decoded part 0 as Kubernetes (lines 1-38)
# decoded part 1 as Kubernetes (lines 40-60)
# decoded part 2 as Kubernetes (lines 62-81)
# decoded part 3 as Kubernetes (lines 83-102)
ok 1 - validating document "httpproxy/001-required-field-validation.yaml"
ok 2 - compiling test document
# hydrated projectcontour.io/v1:HTTPProxy object 'default/missing-condition-header'
ok 3 - hydrating Kubernetes object lines 1-38
ok 4 - matching anonymous Kubernetes object
# performing update operation on HTTPProxy 'default/missing-condition-header'
ok 5 - updating Kubernetes object
# checking update of HTTPProxy 'default/missing-condition-header'
ok 6 - running object update check
# hydrated projectcontour.io/v1:HTTPProxy object 'default/missing-virtualhost-name'
ok 7 - hydrating Kubernetes object lines 40-60
ok 8 - matching anonymous Kubernetes object
# performing update operation on HTTPProxy 'default/missing-virtualhost-name'
ok 9 - updating Kubernetes object
# checking update of HTTPProxy 'default/missing-virtualhost-name'
ok 10 - running object update check
# hydrated projectcontour.io/v1:HTTPProxy object 'default/missing-tcp-services'
ok 11 - hydrating Kubernetes object lines 62-81
ok 12 - matching anonymous Kubernetes object
# performing update operation on HTTPProxy 'default/missing-tcp-services'
ok 13 - updating Kubernetes object
# checking update of HTTPProxy 'default/missing-tcp-services'
ok 14 - running object update check
# hydrated projectcontour.io/v1:HTTPProxy object 'default/missing-includes-name'
ok 15 - hydrating Kubernetes object lines 83-102
ok 16 - matching anonymous Kubernetes object
# performing update operation on HTTPProxy 'default/missing-includes-name'
ok 17 - updating Kubernetes object
# checking update of HTTPProxy 'default/missing-includes-name'
ok 18 - running object update check
1..18
(kind-contour-integration:default)slokas:testsuite/ (jpeach-integration-code-drop) $ ./run-test-case.sh httpproxy/002-header-condition-match.yaml                                                                            [16:02:52]
TAP version 13
# reading document from httpproxy/002-header-condition-match.yaml
# decoding document with 14 parts from httpproxy/002-header-condition-match.yaml
# decoded part 0 as Kubernetes (lines 1-22)
# decoded part 1 as Kubernetes (lines 24-32)
# decoded part 2 as Kubernetes (lines 34-42)
# decoded part 3 as Kubernetes (lines 44-52)
# decoded part 4 as Kubernetes (lines 54-62)
# decoded part 5 as Kubernetes (lines 64-72)
# decoded part 6 as Kubernetes (lines 74-82)
# decoded part 7 as Kubernetes (lines 84-92)
# decoded part 8 as Kubernetes (lines 94-102)
# decoded part 9 as Kubernetes (lines 104-112)
# decoded part 10 as Kubernetes (lines 114-158)
# decoded part 11 as Rego (lines 160-172)
# decoded part 12 as Rego (lines 174-192)
# decoded part 13 as Rego (lines 194-273)
ok 1 - validating document "httpproxy/002-header-condition-match.yaml"
ok 2 - compiling test document
# hydrated apps/v1:Deployment object 'default/echo-header-present'
ok 3 - hydrating Kubernetes object lines 1-22
ok 4 - matching anonymous Kubernetes object
# performing update operation on Deployment 'default/echo-header-present'
ok 5 - updating Kubernetes object
# checking update of Deployment 'default/echo-header-present'
ok 6 - running object update check
# hydrated v1:Service object 'default/echo-header-present'
ok 7 - hydrating Kubernetes object lines 24-32
ok 8 - matching anonymous Kubernetes object
# performing update operation on Service 'default/echo-header-present'
ok 9 - updating Kubernetes object
# checking update of Service 'default/echo-header-present'
ok 10 - running object update check
# hydrated apps/v1:Deployment object 'default/echo-header-contains'
ok 11 - hydrating Kubernetes object lines 34-42
ok 12 - matching anonymous Kubernetes object
# performing update operation on Deployment 'default/echo-header-contains'
ok 13 - updating Kubernetes object
# checking update of Deployment 'default/echo-header-contains'
ok 14 - running object update check
# hydrated v1:Service object 'default/echo-header-contains'
ok 15 - hydrating Kubernetes object lines 44-52
ok 16 - matching anonymous Kubernetes object
# performing update operation on Service 'default/echo-header-contains'
ok 17 - updating Kubernetes object
# checking update of Service 'default/echo-header-contains'
ok 18 - running object update check
# hydrated apps/v1:Deployment object 'default/echo-header-notcontains'
ok 19 - hydrating Kubernetes object lines 54-62
ok 20 - matching anonymous Kubernetes object
# performing update operation on Deployment 'default/echo-header-notcontains'
ok 21 - updating Kubernetes object
# checking update of Deployment 'default/echo-header-notcontains'
ok 22 - running object update check
# hydrated v1:Service object 'default/echo-header-notcontains'
ok 23 - hydrating Kubernetes object lines 64-72
ok 24 - matching anonymous Kubernetes object
# performing update operation on Service 'default/echo-header-notcontains'
ok 25 - updating Kubernetes object
# checking update of Service 'default/echo-header-notcontains'
ok 26 - running object update check
# hydrated apps/v1:Deployment object 'default/echo-header-exact'
ok 27 - hydrating Kubernetes object lines 74-82
ok 28 - matching anonymous Kubernetes object
# performing update operation on Deployment 'default/echo-header-exact'
ok 29 - updating Kubernetes object
# checking update of Deployment 'default/echo-header-exact'
ok 30 - running object update check
# hydrated v1:Service object 'default/echo-header-exact'
ok 31 - hydrating Kubernetes object lines 84-92
ok 32 - matching anonymous Kubernetes object
# performing update operation on Service 'default/echo-header-exact'
ok 33 - updating Kubernetes object
# checking update of Service 'default/echo-header-exact'
ok 34 - running object update check
# hydrated apps/v1:Deployment object 'default/echo-header-notexact'
ok 35 - hydrating Kubernetes object lines 94-102
ok 36 - matching anonymous Kubernetes object
# performing update operation on Deployment 'default/echo-header-notexact'
ok 37 - updating Kubernetes object
# checking update of Deployment 'default/echo-header-notexact'
ok 38 - running object update check
# hydrated v1:Service object 'default/echo-header-notexact'
ok 39 - hydrating Kubernetes object lines 104-112
ok 40 - matching anonymous Kubernetes object
# performing update operation on Service 'default/echo-header-notexact'
ok 41 - updating Kubernetes object
# checking update of Service 'default/echo-header-notexact'
ok 42 - running object update check
# hydrated projectcontour.io/v1:HTTPProxy object 'default/header-conditions'
ok 43 - hydrating Kubernetes object lines 114-158
ok 44 - matching anonymous Kubernetes object
# performing update operation on HTTPProxy 'default/header-conditions'
ok 45 - updating Kubernetes object
# checking update of HTTPProxy 'default/header-conditions'
ok 46 - running object update check
ok 47 - running Rego check lines 160-172
ok 48 - running Rego check lines 174-192
ok 49 - running Rego check lines 194-273
1..49
```

File: _integration/testsuite/fixtures/httpbin.yaml (hunk 1 above)
Comment (jpeach) [reply _integration/testsuite/fixtures/httpbin.yaml:31]: Maybe?

File: _integration/testsuite/README.md (hunk 3 above)
Comment (jpeach) [reply _integration/testsuite/README.md:10]: Yup :)

File: _integration/testsuite/make-kind-cluster.sh (hunk 5 above)
Comment (jpeach) [reply _integration/testsuite/make-kind-cluster.sh:53]: This script is specifically for the case where you want to test what you have in your repository. If you have an existing cluster, you don't need it.

File: _integration/testsuite/README.md (hunk 7 above)
Comment (jpeach) [reply _integration/testsuite/README.md:12]: There were no failures, so this was good. We probably need to think about some kind of summary mode, but I was also thinking that in CI we run each one in a parallel task so they show up as discrete items in the CI task view.

File: _integration/testsuite/httpproxy/004-https-sni-enforcement.yaml (hunk 4 above)
Comment (jpeach) [reply _integration/testsuite/httpproxy/004-https-sni-enforcement.yaml:20]: There's a trade-off between minimizing copy/paste and keeping things obvious, and I mostly haven't minded a bit of extra copy/paste so far. Maybe one way to approach this particular issue is for tests to specify a table of GVRs that they will need, then a library function to do the set difference for the skip query?

File: _integration/testsuite/fixtures/httpbin.yaml (hunk 1 above)
Comment (jpeach) [reply _integration/testsuite/fixtures/httpbin.yaml:31]: For now, I'd like to stick with latest. It also looks like neither of the 2 images we use publish anything except "latest" :)

File: _integration/testsuite/fixtures/ingress-conformance-echo.yaml (hunk 2 above)
Comment (jpeach) [reply _integration/testsuite/fixtures/ingress-conformance-echo.yaml:31]: This only has the "latest" tag

File: _integration/testsuite/httpproxy/004-https-sni-enforcement.yaml (hunk 4 above)
Comment (jpeach) [reply _integration/testsuite/httpproxy/004-https-sni-enforcement.yaml:20]: We could probably reduce this to
```
skip[msg] {
   not resources.is_supported_version("certificates", "cert-manager.io/v1alpha2")
   msg := "cert-manager is not installed"
 }
```

**Comment by jpeach** [comment]
@stevesloka Anything more we should address here?

File: _integration/testsuite/fixtures/httpbin.yaml (hunk 1 above)
Comment (stevesloka) [reply _integration/testsuite/fixtures/httpbin.yaml:31]: Ahh, I see. My feat with these types of tests is that image changes and breaks our tests and we think there's a regression. I'm always leary of using "latest" for anything.

**Pull request merged**